### PR TITLE
a fix for ecPolicyName. also implemented the hierarchical naming func…

### PR DESCRIPTION
**Pull request by riceluxs1t**
Two changes.

1. redundancy_form field of type int is named to ecPolicyName of type String. 
-> In HDFS, the client can register any arbitrary EC policy implemented on the client side with some unique name. To my knowledge, it is the client that knows how to interpret the EC policy name. This policy name is thus returned to the client when it calls the "get_block_locations" rpc call. That said, we must store the given ecPolicyName in string to not lose any information.

2. hierarchical naming scheme implemented.
-> 1) block group ids are pretty much randomly generated 64 bits with its highest bit set to 1 and its lowest 16 bits all set to zeros. 2) storage block ids have the same structure except that its lowest 16 bits represent the index within the block group.

TODO: add unit test cases for the hierarchical naming functions.

### CODE REVIEW

**Comment by riceluxs1t** [comment]
@eddiedugan @LINKIWI not immediately clear what caused Travis to fail. any idea?

**Comment by riceluxs1t** [comment]
Summary of this PR

1. Implemented the hierarchical naming scheme, which is used to generate block group ids and their storage block ids. Using this naming scheme helps with memory efficiency as it eliminates the name node having to manage the reverse mapping from storage block ids to their parent block group ids.

2. Made more ProtoBuff changes. Specifically, the LocatedBlockProto message now contains the storageIDs field. In the case of {EC + striped} files, the response LocatedBlockProto message must contain this field so the client knows what to name each storage block.

3. Must figure out this one question. “How does the name node know what the total number of storage blocks is within a block group?” Depending on which EC encoding scheme is used, the number may vary. 9 for RS(6,3) and 12 for RS(8,4), N+1 for XOR(N), etc. It could be the case that the “replicationFactor” field of AddBlockRequest message is supposed to be interpreted as the total number of storage blocks in EC+striped case. Or it could be that each EC policy and some information about it are also maintained on the NameNode side and is looked up whenever necessary. Currently, the logic that figure this out is encapsulated into its own function and must be fully figured out later.

4. The question of what metadata changes must occur has been figured out. 1) The file tree side of ZooKeeper stays the same. A {EC+striped} file has children nodes that corresponds to block group ids. (analogous to how a {Replication+contiguous} file has children nodes that correspond to logical blocks. 2) The /block_group_locations/ tree will have one level of nodes that correspond to block groups and their children level of nodes that correspond to storage blocks of each block group.

5. In the process of implementing “add_block_group” helper function. This helper function will be called whenever a client wants to add a new logical block (in other words, a new block group) to a file. What this function does is that it 1) comes up with the new block group id. 2) comes up with storage block ids using the hierarchical naming scheme. 3) performs metadata changes required on the ZooKeeper side. The details of part 3 are being implemented.

6. Unit test coverage for task 1 above must be added.

**Comment by LINKIWI** [comment]
@riceluxs1t Seems like the Hadoop mirror is down. Thanks for reporting, a fix is incoming.

**Comment by LINKIWI** [comment]
@riceluxs1t Please rebase against `origin/develop`, the issue has been fixed. Note that this isn't merge-blocking since this PR isn't for `develop`.

File: zookeeper/source/zk_nn_client.cc [hunk 1]
```diff
@@ -1322,14 +1317,70 @@ bool ZkNnClient::add_block(const std::string &file_path,
...
+    u_int64_t res = 1ull << 63;  // filled with 1 and 63 zeros.
+    res = res & block_group_id;  // & to fill bit2~bit48.
+    res = res & index_within_group;  // & to fill the lower 16 bits.
+    return res;
 }
 
+u_int64_t ZkNnClient::generate_block_group_id() {
+    u_int64_t res;
+    util::generate_uuid(res);  // generate some random 64 bits.
+    res = res | (1ull << 63);  // filp the highest bit to one.
```
Comment (willkoh) [inline zookeeper/source/zk_nn_client.cc]: Why is this here? Does this bit always have be 1 regardless of whether it's EC or Replication?

Comment (willkoh) [reply zookeeper/source/zk_nn_client.cc]: Or wait, is this function only called for EC? Since contiguous won't need a block_group_id?

**Review by willkoh — APPROVED** [review]
It seems like you took out the normal blockID generation mechanisms and put in the EC blockID generation methods. But, if it's replication, does it still have a way to get a blockID? It might be somewhere in the code that you didn't touch, and I'm just not seeing it. Other than that, looks good.

File: zookeeper/source/zk_nn_client.cc (hunk 1 above)
Comment (adamnsm1) [reply zookeeper/source/zk_nn_client.cc]: Yeah, this function is only for EC files

File: zookeeper/source/zk_nn_client.cc [hunk 2]
```diff
@@ -1322,14 +1317,70 @@ bool ZkNnClient::add_block(const std::string &file_path,
...
+    return storage_block_id & mask;
+}
+
+
+uint32_t ZkNnClient::get_total_num_storage_blocks(
+        const std::string &fileName,
+        u_int64_t &block_group_id) {
+    // TODO(nate): actually figure out where this information is stored.
+    // TODO(nate): is it supposed to be figured out from ecPolicyName?
+    return 9; // arbitrarily assume RS(6, 3)
```
Comment (adamnsm1) [inline zookeeper/source/zk_nn_client.cc]: I think this is a fine implementation for now. Our first iteration of EC can enforce RS(6, 3) and add the ability to specify later.

**Review by adamnsm1 — APPROVED** [review]
Everything mostly looks good, good work Nate!

File: zookeeper/source/zk_nn_client.cc (hunk 1 above)
Comment (riceluxs1t) [reply zookeeper/source/zk_nn_client.cc]: @willkoh util::generate_uuid is the function currently used to generate a new blockID

**Comment by riceluxs1t** [comment]
@LINKIWI just to be clear, rebase what branch against origin/develop (which I think you pushed a fix to) ? EC branch or my feature branch? I feel like rebasing either branch against the origin/develop branch at this point will introduce many merge conflicts. help lol

**Comment by riceluxs1t** [comment]
Just for the record, all the changes this PR introduced are good. The travis failure was due to a hadoop mirror failure.
